extensions/sql-database-projects/src/dialogs/addDatabaseReferenceQuickpick.ts

```diff
@@ -0,0 +1,276 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import path = require('path');
+import * as vscode from 'vscode';
+import * as constants from '../common/constants';
+import { getSqlProjectsInWorkspace, isValidSqlCmdVariableName, removeSqlCmdVariableFormatting } from '../common/utils';
+import { AddDatabaseReferenceSettings } from '../controllers/projectController';
+import { IDacpacReferenceSettings, IProjectReferenceSettings, ISystemDatabaseReferenceSettings } from '../models/IDatabaseReferenceSettings';
+import { Project } from '../models/project';
+import { getSystemDatabase, getSystemDbOptions, promptDacpacLocation } from './addDatabaseReferenceDialog';
+
+interface DbServerValues {
+	dbName?: string,
+	dbVariable?: string,
+	serverName?: string,
+	serverVariable?: string
+}
+
+/**
+ * Create flow for adding a database reference using only VS Code-native APIs such as QuickPick
+ * @param connectionInfo Optional connection info to use instead of prompting the user for a connection
+ */
+export async function addDatabaseReferenceQuickpick(project: Project): Promise<AddDatabaseReferenceSettings | undefined> {
+
+	const otherProjectsInWorkspace = (await getSqlProjectsInWorkspace()).filter(p => p.fsPath !== project.projectFilePath);
+
+	// 1. Prompt for reference type
+	// Only show project option if we have at least one other project in the workspace
+	const referenceTypes = otherProjectsInWorkspace.length > 0 ?
+		[constants.projectLabel, constants.systemDatabase, constants.dacpacText] :
+		[constants.systemDatabase, constants.dacpacText];
+
+	const referenceType = await vscode.window.showQuickPick(
+		referenceTypes,
+		{ title: constants.referenceType, ignoreFocusOut: true });
+	if (!referenceType) {
+		// User cancelled
+		return undefined;
+	}
+
+	switch (referenceType) {
```

Comment:
Maybe add a comment saying that the steps are continued in these methods to avoid confusion.